modules/programs/brew-bundle.nix

```diff
@@ -0,0 +1,213 @@
+
+let
+  cfg = config.programs.brew-bundle;
+
+  brewfileSection = heading: type: entries:
+    if entries != [] then
+      "# ${heading}\n" + (concatMapStrings (name: "${type} \"${name}\"\n") entries) + "\n"
+    else "";
+
+  masBrewfileSection = entries:
+    if entries != {} then
+      "# Mac App Store apps\n" +
+      concatStringsSep "\n" (mapAttrsToList (name: id: ''mas "${name}", id: ${toString id}'') entries) +
+      "\n"
+    else "";
+
+  brewfile = pkgs.writeText "Brewfile" (
+    (brewfileSection "Taps" "tap" cfg.taps) +
+    (brewfileSection "Brews" "brew" cfg.brews) +
+    (brewfileSection "Casks" "cask" cfg.casks) +
+    (masBrewfileSection cfg.masApps) +
+    (brewfileSection "Docker contrainers" "whalebrew" cfg.whalebrews) +
+    (if cfg.extraConfig != "" then "# Extra config\n" + cfg.extraConfig else "")
+  );
+
+  brew-bunble-options =
+    "--file='${brewfile}' --no-lock" +
+    (if cfg.cleanupType == "uninstall" || cfg.cleanupType == "zap" then " --cleanup" else "") +
+    (if cfg.cleanupType == "zap" then " --zap" else "");
+
+in
+
+{
+  options.programs.brew-bundle = {
+    enable = mkEnableOption ''
+      configuring your Brewfile, and installing/updating the formulas therein via
+      the <command>brew bundle</command> command, using <command>nix-darwin</command>.
+
+      Note that enabling this option does not install Homebrew. See the Homebrew website for
+      installation instructions: https://brew.sh
+    '';
+
+    cleanupType = mkOption {
+      type = with types; enum [ "none" "uninstall" "zap" ];
+      default = "none";
+      example = "uninstall";
+      description = ''
+        This option manages what happens to formulas installed by Homebrew, that aren't present in
+        the Brewfile generated by this module.
+
+        When set to <literal>"none"</literal> (the default), formulas not present in the generated
+        Brewfile are left installed.
+
+        When set to <literal>"uninstall"</literal>, <command>nix-darwin</command> invokes
+        <command>brew bundle install</command> with the <command>--cleanup</command> flag. This
+        uninstalls all formulas not listed in generate Brewfile, i.e.,
+        <command>brew uninstall</command> is run for those formulas.
+
+        When set to <literal>"zap"</literal>, <command>nix-darwin</command> invokes
+        <command>brew bundle install</command> with the <command>--cleanup --zap</command>
+        flags. This uninstalls all forumalas not listed in the generated Brewfile, and if the
+        formula is a cask, removes all files associated with the cask. In other words,
+        <command>brew uninstall --zap</command> is run for all those formulas.
+
+        If you plan on exclusively using <command>nix-darwin</command> to manage formulas installed
+        by Homebrew, you probably want to set this option to <literal>"uninstall"</literal> or
+        <literal>"zap"</literal>.
+      '';
+    };
+
+    setNoLockEnvvar = mkOption {
```

Comment:
That it is an environment variable is an implementation detail, IMHO.  And also "set" signals a kind of imperative rather than declarative intent.  I think `lockfile`.

Reply:
Good point about using the word "set" here, and the fact that the name is probably to anchored the the implementation details.

One thing I'm trying to track with the options in this module is that some effect what happens when the activation script runs, while others (like this one) effect what happens when you run the `brew bundle` command yourself. During activation a lockfile is never generated, and so it would be nice if the option name could make it clear somehow that this  effects how `brew bundle` works when the user runs it. Maybe it's just to hard/cumbersome to communicate that in the name, and so the description will have to suffice.

Reply:
Alright, I've changed this option to from `setNoLockEnvvar` to `userConfig.noLock`, which I don't love, but I think is an improvement. (Did something similar for `setBrewfileEnvbar` which is now `userConfig.brewfile`.)

Reply:
Resolving this comment thread, and picking the conversation up about these options, in a new thread.